File: Reqnroll/Chapter15/15.6-PolymorphicDriverPattern/After/WIMP.Specs/Drivers/AuthenticationServiceDriver.cs
using WIMP.App.RestApi;
using WIMP.App.Services;
using WIMP.Specs.Support;

namespace WIMP.Specs.Drivers;

public class AuthenticationServiceDriver(AuthenticationService authService) : IAuthenticationDriver
{
    public TestAction<LoginResponse> Login(string customerName, string password) =>
        new LambdaAction<LoginResponse>("Login", () =>
        {
            var result = authService.Login(customerName, password);
            return result.Successful ? new LoginResponse(result.Value, customerName) :
                    throw new WimpActionFailedException(result.ErrorMessage);
        });

    public TestAction<VoidReturn> Register(string customerName, string email) =>
        new LambdaAction("Register", () =>
        {
            var result = authService.Register(customerName, email);
            if (!result.Successful)
            {
                throw new WimpActionFailedException(result.ErrorMessage);
            }
        });
}

File: CucumberJVM/Chapter15/15.6-PolymorphicDriverPattern/After/src/test/java/com/wimp/app/specs/drivers/AuthenticationServiceDriver.java
package com.wimp.app.specs.drivers;

import com.wimp.app.models.ServiceResult;
import com.wimp.app.restapi.LoginResponse;
import com.wimp.app.services.AuthenticationService;
import com.wimp.app.specs.support.LambdaAction;
import com.wimp.app.specs.support.TestAction;
import com.wimp.app.specs.support.TestActionFailedException;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.stereotype.Component;

@Component
@ConditionalOnProperty(name = "test.execution.test-target", havingValue = "service-api")
public class AuthenticationServiceDriver implements AuthenticationDriver {
    private final AuthenticationService authenticationService;

    public AuthenticationServiceDriver(AuthenticationService authenticationService) {
        this.authenticationService = authenticationService;
    }

    @Override
    public TestAction<LoginResponse> login(String customerName, String password) {
        return new LambdaAction<>("Login", "%s/%s".formatted(customerName, password), () -> {
            ServiceResult<String> result = authenticationService.login(customerName, password);
            if (!result.successful()) {
                throw new TestActionFailedException(result.errorMessage());
            }
            return new LoginResponse(result.value(), customerName);
        });
    }
}

